FixedLayer: Calculate top: auto by mutating bottom

jridgewell opened the pull request:
This uses a new approach to calculate if a fixed-position (or sticky-position) element is `top: auto` by mutating the `bottom` to a (very large) value. This should detect autos in any circumstance, fixing any false positives that were caught (causing bad transfers that are later removed). This is the root cause behind https://github.com/ampproject/amphtml/issues/10343 and https://github.com/ampproject/amphtml/issues/10413.

Fixes https://github.com/ampproject/amphtml/issues/10343
Fixes https://github.com/ampproject/amphtml/issues/10413.

Note that #10343 will still require an explicit `top: 0px` to transfer without flickering.

jridgewell commented:
/to @dvoytenko

dvoytenko commented on src/service/fixed-layer.js:
```diff
@@ -241,74 +241,66 @@ export class FixedLayer {
-          } else {
-            setStyle(el, 'top', 'auto');
-          }
+          const element = elements[i].element;
```
Can you up to document all steps that take places in bulks. E.g. "Change bottom -> remeasure top -> revert bottom"  and explain the result

jridgewell replied:
Will do.

dvoytenko commented on src/service/fixed-layer.js:
```diff
@@ -241,74 +241,66 @@ export class FixedLayer {
+            bottom,
+            opacity,
+            zIndex,
+            transform,
```
Weird that the old code doesn't try to resolve vendor prefixing for `transform`. Where do we still have that problem? IE?

jridgewell replied:
I think it's actually old Safari and old Android that we did prefixes for. I'll update.

dvoytenko commented on src/service/fixed-layer.js:
```diff
@@ -320,28 +312,15 @@ export class FixedLayer {
           }
 
-          const bottom = style.bottom;
-          const opacity = parseFloat(style.opacity);
```
Typically we try to stay away from numerical comparisons with implicit cast. Are we scrapping that?

jridgewell replied:
Whoops, that's an an error. We still need to parse here.

dvoytenko commented on src/service/fixed-layer.js:
```diff
@@ -557,9 +536,18 @@ export class FixedLayer {
     // paddingTop.
-    if (state.top && (state.fixed || (state.sticky && !this.transfer_))) {
-      setStyle(element, 'top', `calc(${state.top} + ${this.paddingTop_}px)`);
+    // TODO
```
What's the `TODO`?

jridgewell replied:
I use TODOs as markers for while I'm coding, then forget to remove them.

dvoytenko commented on src/service/fixed-layer.js:
```diff
@@ -557,9 +536,18 @@ export class FixedLayer {
 
     // Update `top`. This is necessary to adjust position to the viewer's
     // paddingTop.
-    if (state.top && (state.fixed || (state.sticky && !this.transfer_))) {
```
I believe both `fixed` and `sticky` could be false. Is it valid to remove it from this expression in the new code? It might have some undesired affect inside the `position:relative` parent...

jridgewell replied:
I'm not checking for `state.top`, which can only be truthy if `fixed || sticky`. I can be more explicit if you'd like.

dvoytenko replied:
Either more explicit or comment above that `state.top` is only truthy when fixed or sticky - it's harder to read spread out state so some overprotection is warranted.

dvoytenko commented on src/service/fixed-layer.js:
```diff
@@ -557,9 +536,18 @@ export class FixedLayer {
+      if (state.fixed || !this.transfer_) {
+        setStyle(element, 'top', `calc(${state.top} + ${this.paddingTop_}px)`);
+      } else if (this.committedPaddingTop_ === this.paddingTop_) {
+        setStyle(element, 'top', state.top);
```
Ditto: why do we need to update `top` styles of non-fixed/non-sticky elements? Same is with `else` below?

jridgewell replied:
To offset for iOS's `padding-top`. Without these lines, we get fixed and stickies that remain `54px` off the top of the viewport. This case is explicitly for when the header is showing (in which case, adding 54px would be compounding the 54px `padding-top`). The next case is for when the header is hidden, in which case we need to subtract that 54px.

jridgewell commented:
Ok, I've addressed everything. I've also tested (in a prod viewer that hides/shows the header) in iOS Safari, iOS Chrome, and Android Chrome, and everything is working.

This does not address `position: fixed`s that are both `top: auto` and `bottom: auto`, though. On iOS, the won't get the negative offsets to offset the header (when it's hidden). We could extend the the `top`/`bottom` testing to detect `bottom: auto`s too, and apply the offset just like we do for `top: auto`.

Also, why don't we lift the `top: 0px`/`bottom: 0px` restriction for transfer? We should be able to detect any specified distance now.

dvoytenko commented:
LGTM with one request on more explicit or documented sticky/fixed.

> Also, why don't we lift the top: 0px/bottom: 0px restriction for transfer?
We can try/do in a separate PR. TBH, the operation is so heavy that I initially decided to restrict it to couple very specific use cases. Namely header and footer. Also, the iOS bug is most noticable for header and footer. But we can certainly expand if you are feeling good about it.
